### ISSUE TITLE
Moving the cursor to another screen doesn't always cause that screen to focus

### ISSUE THREAD

**Issue by ParaSait** [qtile/qtile#2657]
# Issue description
When I move the mouse cursor to another screen, the behavior that I expect is that that screen always becomes focused. However, the screen will now only become focused when you move it onto a window that's on it, not when you move it onto a blank area or a bar. Clicking on the blank area or bar doesn't trigger the focus either.
(By "focused", I mean the screen that any commands apply to, such as spawning a new window.)

# Qtile version
v0.18.0 (in v0.17.0, this was still working)

**Correction:** actually, in 0.17.0, it seems only clicking will focus it, not just moving the mouse over...

# Configuration
I've reproduced it with my own configuration and with the default fallback configuration, so I don't believe it's related to it.

**Comment by yobleck** [comment]
Looks similar to issue #2098
Personally I'd be fine with a qtile constantly watching the root window. Could be a config option with a warning about increased cpu usage

**Comment by ParaSait** [comment]
I'd just like to see the v0.17.0 functionality repaired, i.e., focus on the window & its screen when hovering the mouse over a window, and focus on the screen when clicking on a blank space on the screen. I got it wrong when I implied that just hovering a blank space would change the focus in v0.17.0, but the fact remains that clicking blank space doesn't do anything in v0.18.0, which I believe must be a bug.
Of course, splitting the `follow_mouse_focus` option into two separate options (or something to that effect) would be nice to have.

**Comment by m-col** [comment]
> Personally I'd be fine with a qtile constantly watching the root window. Could be a config option with a warning about increased cpu usage

I'd be in favour of this too, as it is consistent with `follow_mouse_focus` but with screens. I don't think it would affect CPU usage noticably and would be pretty easy to add in if anybody wants to look into it.

**Comment by jas3333** [comment]
I'm having the same issue. Windows kept opening on my main monitor even after clicking on the second screen. Was driving me crazy so I went back to version 0.17.

**Comment by yobleck** [comment]
It's taking me awhile to look into this but there's a temporary work around with the CurrentScreen widget.
put:
`widget.CurrentScreen(mouse_callbacks={"Button1": lambda: qtile.focus_screen(0)}),`
in the bar on every screen (replace `0` with the screen number). This requires you to move the mouse to the bar as you switch screens and click the widget, which is a hassle but serviceable

**Comment by yobleck** [comment]
Another janky workaround:
Add:
```
try:
    if self.qtile.config.follow_mouse_focus:
        index_under_mouse = self.qtile.find_screen(event.root_x, event.root_y).index
        if index_under_mouse != self.qtile.current_screen.index:
            self.qtile.focus_screen(index_under_mouse)
except:
    pass
```
to [line 371 of libqtile/backend/x11/core.py](https://github.com/qtile/qtile/blob/2f4722bb30e39510e985a08da1d28dbdd6ccd232/libqtile/backend/x11/core.py#L371).
Unfortunately this only fires when going from a window or bar to background. Not when going from background to background on different screens. This results in some mouse waving (usually over the bar) to get the code to trigger.

I don't think this is a permanent though because there is no screen to screen X event, and I'm not sure how we would implement one.

Is there a better place this code could be placed so that it's running constantly and not just when an X event fires?

**Comment by m-col** [comment]
Perhaps to the top of `Qtile.process_button_motion` in core/manager.py? Both the X and Wayland backends call that method when there is cursor motion and provide the cursor coordinates.

**Comment by yobleck** [comment]
> Perhaps to the top of `Qtile.process_button_motion` in core/manager.py? Both the X and Wayland backends call that method when there is cursor motion and provide the cursor coordinates.

Unfortunately that function is triggered by MotionNotify X events and the only way I could trigger that event was by dragging
a floating window around

**Comment by m-col** [comment]
How were you checking? The first couple lines of that method make it return early when a drag is not ongoing, implying that it does get called even when there is not a drag.

**Comment by yobleck** [comment]
I chucked my code up top before `if self._drag`
I also put a logger in backend/x11/core.Core._get_target_chain to print out all events. The only time it showed MotionNotify was with dragging.

**Comment by m-col** [comment]
Ah I think I know why - the events are only emitted by the backend when 
the right combination of modifiers has been grabbed and is held. We'd 
need to essentially replicate what `Drag` does WRT grabbing but with no 
modifiers so that they are always emitted when there is movement 
(without any mods).

**Comment by yobleck** [comment]
I'm just thowing another idea out there:
The idea of a transparent background window was previously [mentioned](https://github.com/qtile/qtile/issues/2098#issuecomment-760226530),
and I just remembered that KDE uses a window called `Desktop — Plasma` to display the background image.
Programs like feh and imagemagick draw on the root window and the KDE window has to be killed to to show them.
What if the KDE devs encountered similar issues?

So what if we drew the background image (blank black by default) as internal windows (one for each screen) and just had everything draw over it? (not sure how this would work with bar gaps)
That way when the mouse moves off a window it moves onto the background window which should trigger an X event.
Combine this with a handler function that checks for the background window ids and change screen focus accordingly.

**Comment by yobleck** [comment]
second half of janky solution mentioned [here](https://github.com/qtile/qtile/issues/2657#issuecomment-890562210)

in config do:
```
from libqtile import qtile
@hook.subscribe.startup
def some_func():
    internal_window(qtile)

def internal_window(qtile):
    #replace x,y,width,height with your values (subtract bar height from screen res)
    win1 = qtile.core.create_internal(x=0,y=0,width=800,height=876)
    #name makes window easier to find and kill from qtile shell internal_windows()
    win1.name = "internal_background_1"
    win1.place(0,0,800,876,0,"#000000")
    win1.unhide()
    win2 = qtile.core.create_internal(x=800,y=0,width=800,height=876)
    win2.name = "internal_background_2"
    win2.place(800,0,800,876,0,"#000000")
    win2.unhide()
```
This creates some internal windows that take on the appearance of the desktop background
and ensure that the code in the previous comment triggers when crossing screen borders.

Warning: sometimes windows get stuck behind the internal window and have to be moved to other groups or killed.

**Comment by m-col** [comment]
Hmm, not sure this is a good idea because it might lead to unexpected side effects, such as giving misleading output when you xprop the root window. Also, without proper Z axis handling clicking it will raise it above all the windows and then you're stuck. It will also conflict with desktop manager windows (such as the one plasma uses; there are standalone versions people might use).

An alternative would be to use a pointer barrier at the edges of screens and listen to their events. They are in the xfixes extension.

**Comment by yobleck** [comment]
I'm having trouble figuring out how to implement pointer barriers (X11 documentation is a little sparse and I can't find any examples), but it turns out it may not be necessary (no invisible background windows either :) ).

When I moved my code over from xephyr to my main install most of the issues I was having went away (xephyrs multi-monitor support is lacking).
Turns out having:
```
try:
    if self.qtile.config.follow_mouse_focus and not self.qtile.config.cursor_warp:
        if hasattr(event, "root_x") and hasattr(event, "root_y"):
            screen = self.qtile.find_screen(event.root_x, event.root_y)
            if screen:
                index_under_mouse = screen.index
                if index_under_mouse != self.qtile.current_screen.index:
                    self.qtile.focus_screen(index_under_mouse, warp=False)
except Exception as e:
    print("DIDNT WORK", e) #replace with logging later
```
at line 371 of [backend/x11/core.py](https://github.com/qtile/qtile/blob/01816d7d5b97de4c9239149ff5b6b3fbe93c10f6/libqtile/backend/x11/core.py#L371)
seems to completely fix the issue.

It still feels kind of janky to me though. I had to add all the if statements to deal with various different errors, and it seems like the code has to be here(or maybe in _xpoll) rather than in a handler function because then it runs when ANY Xevent is processed. 

If there is some way to refine this so that it only catches the relevant events EnterNotify, LeaveNotify, ConfigureNotify (This one is particularly important because we can catch windows before they are drawn/placed/mapped and change screens just in the nick of time(but for some reason it doesn't work in xephyr))  and MotionNotify I'd like to hear people's thoughts.

side note: This seems to be incompatible with the config option warp_cursor, causing test/backend/x11/test_window.test_cursor_warp() to fail if warp_cursor=True and the code is run. But I don't think it's a big issue because if the user is manually moving their mouse to change focus then warping the cursor from keyboard inputs is probably not in their work flow. Someone correct me if I'm wrong.

**Comment by GRFreire** [comment]
Doing some tests I found the commit that caused this issue. 67f97604
I've tried to change some stuff but I am not familiar with the code base, so I'm sharing it here.

It's not a big commit, so it should be quite easy to fix (idk, i'm just guessing)

**Comment by yobleck** [comment]
@m-col sorry for not paying enough attention to:

> Ah I think I know why - the events are only emitted by the backend when the right combination of modifiers has been grabbed and is held. We'd need to essentially replicate what `Drag` does WRT grabbing but with no modifiers so that they are always emitted when there is movement (without any mods).

Turns out it's super easy to enable event listening for the root window.
by adding `EventMask.ButtonPress` to here: https://github.com/qtile/qtile/blob/8454c6294d276c8fffcf330a5236b10920e146a3/libqtile/backend/x11/core.py#L109-L116
this pre-existing code:
https://github.com/qtile/qtile/blob/8454c6294d276c8fffcf330a5236b10920e146a3/libqtile/backend/x11/core.py#L771-L775
starts working. This gets us back to the way it was before (0.17) where clicking the root window focuses that screen.

The questions is whether or not to enable `EventMask.PointerMotion` on the root window and then add my code to where you suggested here:
> Perhaps to the top of `Qtile.process_button_motion` in core/manager.py? Both the X and Wayland backends call that method when there is cursor motion and provide the cursor coordinates.

This allows pointer motion to change screens but it may slightly increase CPU usage and it completely spams logs in debug mode. My logs are completely full of
```
2021-08-04 18:24:11,562 libqtile core.py:_xpoll():L299  Handling: MotionNotify
2021-08-04 18:24:11,570 libqtile core.py:_xpoll():L296  MotionNotify
```

@ParaSait @jas3333 @GRFreire do you folks want mouse movements to change screen or are you fine with just clicking on the screen?

**Comment by ParaSait** [comment]
Clicking the screen would be fine for me; that's how it worked in the previous version.

**Comment by GRFreire** [comment]
That would be perfect for me.
The default behavior in 0.17 was clicking (i think), not moving the mouse, thus I wont even need to change my workflow.

**Comment by jas3333** [comment]
Clicking the screen would be fine for me as well.

**Comment by m-col** [comment]
I wonder if this is what these old lines that were removed were doing (from core/manager.py):
```py
        # It fixes problems with focus when clicking windows of some specific clients like xterm
        def noop(qtile):
            pass
        self.config.mouse += (Click([], "Button1", lazy.function(noop), focus="after"),)
```
The solution in #2687 is definitely the best!

**Comment by elParaguayo** [comment]
> I wonder if this is what these old lines that were removed were doing (from core/manager.py):
> 
> ```python
>         # It fixes problems with focus when clicking windows of some specific clients like xterm
>         def noop(qtile):
>             pass
>         self.config.mouse += (Click([], "Button1", lazy.function(noop), focus="after"),)
> ```
> 
> The solution in #2687 is definitely the best!

But didn't those lines cause the bug in Chrome and other apps?

**Comment by m-col** [comment]
Yeah I think it was problematic, but we were getting the `ButtonPress` 
events by manually grabbing for the clicks, which is why `ButtonPress` 
was absent from our root window's event mask. The solution is putting 
it in the event mask. Oh.... i see what you're saying. Are we going to 
get those issues again? Ha. In that case...

Would any of you guys be happy to run the changes from the PR and 
ensure that it isn't introducing any strange issues, particularly with 
focussing around clicks, and particularly with chromium and pals?

**Comment by yobleck** [comment]
some quick chromium tests with drop down menus didn't have any issues for me, but it was a very quick test.

**Comment by GRFreire** [comment]
I used to have issues with that with drop down menus in google chrome and vscode scrollbar, none of these bugs are present in the PR branch. Did some tests and all fine to me.

**Comment by m-col** [comment]
Brilliant thank you for checking

**Comment by m-col** [comment]
Closed by #2687

**Comment by yobleck** [comment]
In case anyone wants the mouse movement changes desktop option (uses more cpu and spams logs), 
I made it into a somewhat janky config function:
```
from libqtile import qtile
from xcffib.xproto import EventMask
def mouse_move(qtile):
    qtile.core._root.set_attribute(eventmask=(EventMask.StructureNotify
                                            | EventMask.SubstructureNotify
                                            | EventMask.SubstructureRedirect
                                            | EventMask.EnterWindow
                                            | EventMask.LeaveWindow
                                            | EventMask.ButtonPress
                                            | EventMask.PointerMotion))
    def screen_change(event):
        assert qtile is not None
        if qtile.config.follow_mouse_focus and not qtile.config.cursor_warp:
            if hasattr(event, "root_x") and hasattr(event, "root_y"):
                screen = qtile.find_screen(event.root_x, event.root_y)
                if screen:
                    index_under_mouse = screen.index
                    if index_under_mouse != qtile.current_screen.index:
                        qtile.focus_screen(index_under_mouse, warp=False)
        qtile.process_button_motion(event.event_x, event.event_y)
    setattr(qtile.core, "handle_MotionNotify", screen_change)

@hook.subscribe.startup_once
def my_func_name():
    mouse_move(qtile)
```
